File: pkg/triple_endpoint/2client_spend_tx.go
package triple_endpoint

import (
	"encoding/hex"
	"fmt"
	"log"

	libs "github.com/spycat55/KeymasterMultisigPool/pkg/libs"
	multisig "github.com/spycat55/KeymasterMultisigPool/pkg/libs"

	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
	tx "github.com/bsv-blockchain/go-sdk/transaction"
	sighash "github.com/bsv-blockchain/go-sdk/transaction/sighash"
	"github.com/bsv-blockchain/go-sdk/transaction/template/p2pkh"
)

// 多签 to client，server 提供金额
func SubBuildTripleFeePoolSpendTX(
	prevTxId string,
	serverValue uint64, // server 提供金额
	// cmdValue uint64, // cmd 提供金额
	endHeight uint32, // 区块高度
	serverPublicKey *ec.PublicKey,
	aPrivateKey *ec.PrivateKey,
	bPublicKey *ec.PublicKey,
	isMain bool,
	feeRate float64,
) (*tx.Transaction, uint64, error) {
	aAddress, err := libs.GetAddressFromPublicKey(aPrivateKey.PubKey(), isMain)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to get client address: %w", err)
	}
	bAddress, err := libs.GetAddressFromPublicKey(bPublicKey, isMain)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to get server address: %w", err)
	}

	// 生成公钥
	aPublicKey := aPrivateKey.PubKey()

	transactionTwo := tx.NewTransaction()
	transactionTwo.LockTime = endHeight

	// 创建初始交易的锁定脚本
	prevMultisigScript, err := multisig.Lock([]*ec.PublicKey{serverPublicKey, aPublicKey, bPublicKey}, 2)
	fmt.Printf("================================= prevMultisigScript: %s\n", prevMultisigScript.ToASM())
	if err != nil {
		return nil, 0, fmt.Errorf("failed to create server locking script: %w", err)
	}
	prevMultisigTxLockingAsm := hex.EncodeToString(prevMultisigScript.Bytes())

	sigHash := sighash.Flag(sighash.ForkID | sighash.All)
	aMultisigUnlockingScriptTemplate, err := multisig.Unlock([]*ec.PrivateKey{}, []*ec.PublicKey{serverPublicKey, aPublicKey, bPublicKey}, 2, &sigHash)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to create unlocking script template: %w", err)
	}

	// 添加所有UTXO作为输入
	err = transactionTwo.AddInputFrom(
		prevTxId,
		0,
		prevMultisigTxLockingAsm,
		serverValue,
		aMultisigUnlockingScriptTemplate,
	)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to add input: %w", err)
	}
	transactionTwo.Inputs[0].SequenceNumber = 1

	// 服务器找零脚本
	// serverAddress, err := script.NewAddressFromPublicKey(serverPublicKey, isMain)
	// if err != nil {
	// 	return nil, fmt.Errorf("无法从公钥生成地址: %v", err)
	// }
	serverChangeScript, err := p2pkh.Lock(bAddress)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to create change locking script: %w", err)
	}

	// 添加服务器输出
	transactionTwo.AddOutput(&tx.TransactionOutput{
		Satoshis:      0,
		LockingScript: serverChangeScript,
	})

	// 客户端找零脚本
	// clientAddress, err := script.NewAddressFromPublicKey(clientPublicKey, isMain)
	// if err != nil {
	// 	return nil, fmt.Errorf("failed to create change locking script: %w", err)
	// }
	clientChangeScript, err := p2pkh.Lock(aAddress)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to create change locking script: %w", err)
	}

	// 添加客户端输出
	transactionTwo.AddOutput(&tx.TransactionOutput{
		Satoshis:      serverValue,
		LockingScript: clientChangeScript,
	})

	// 做一个假的签名script，方便计算 size
	unlockingScript, err := multisig.FakeSign(2)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to sign input %d: %w", 1, err)
	}
	transactionTwo.Inputs[0].UnlockingScript = unlockingScript

	// 计算交易大小（字节）
	txSize := transactionTwo.Size()

	// 基于大小计算费用（向上取整到最接近的KB）
	fee := uint64(float64(txSize) / 1000.0 * feeRate)
	if serverValue < fee {
		return nil, 0, fmt.Errorf("not enough balance, need %d, have %d", fee, serverValue)
	}
	if fee == 0 {
		fee = 1
	}

	// 更新找零输出的金额
	transactionTwo.Outputs[1].Satoshis = serverValue - fee

	// transactionTwo.Inputs[0].UnlockingScript = serverSignByte

	return transactionTwo, serverValue - fee, nil
}

func SpendTXTripleFeePoolASign(
	B_Tx *tx.Transaction,
	targetAmount uint64,
	serverPublicKey *ec.PublicKey,
	aPrivKey *ec.PrivateKey,
	bPublicKey *ec.PublicKey,
) (*[]byte, error) {
	// 创建优先级脚本
	priorityScript, err := multisig.Lock([]*ec.PublicKey{serverPublicKey, aPrivKey.PubKey(), bPublicKey}, 2)
	if err != nil {
		return nil, fmt.Errorf("创建优先级脚本失败: %v", err)
	}

	// 设置输入的锁定脚本
	B_Tx.Inputs[0].SetSourceTxOutput(
		&tx.TransactionOutput{
			Satoshis:      targetAmount,
			LockingScript: priorityScript,
		},
	)

	// unlocking script
	sighash := sighash.Flag(sighash.ForkID | sighash.All)
	aMultisigUnlockingScriptTemplate, err := multisig.Unlock([]*ec.PrivateKey{aPrivKey}, []*ec.PublicKey{serverPublicKey, aPrivKey.PubKey(), bPublicKey}, 2, &sighash)
	if err != nil {
		return nil, fmt.Errorf("创建解锁脚本失败: %v", err)
	}

	// 重新签名所有输入
	aSignByte, err := aMultisigUnlockingScriptTemplate.SignOne(B_Tx, 0, aPrivKey)
	if err != nil {
		return nil, fmt.Errorf("a 重新签名输入 %d 失败: %v", 1, err)
	}
	return aSignByte, nil
}

// 构建双端费用池花费交易
// 发起者 utxos, 服务器提供金额， 发起者私钥， 服务器地址
// fee 是 server 提供，我只负责精确的金额
func BuildTripleFeePoolSpendTX(
	A_Tx *tx.Transaction,
	serverValue uint64, // 服务器提供金额
	endHeight uint32, // 区块高度
	serverPublicKey *ec.PublicKey,
	aPrivateKey *ec.PrivateKey,
	bPublicKey *ec.PublicKey,
	isMain bool,
	feeRate float64,
) (*tx.Transaction, *[]byte, uint64, error) {

	txTwo, amount, err := SubBuildTripleFeePoolSpendTX(A_Tx.TxID().String(), serverValue, endHeight, serverPublicKey, aPrivateKey, bPublicKey, isMain, feeRate)
	if err != nil {
		log.Printf("BuildOneB error: %v", err)
		return nil, nil, 0, err
	}

	// log.Printf("------------------------------- BuildOneB success: %v", txTwo.Hex())

	// 重新签名
	clientSignByte, err := SpendTXTripleFeePoolASign(txTwo, serverValue, serverPublicKey, aPrivateKey, bPublicKey)
	if err != nil {
		log.Printf("BuildOneC error: %v", err)
		return nil, nil, 0, err
	}

	return txTwo, clientSignByte, amount, nil
}

File: src/triple_endpoint/2client_spend_tx.ts
import { PrivateKey, PublicKey } from '@bsv/sdk/primitives';
// import Script from '@bsv/sdk/script/Script';
import Transaction from '@bsv/sdk/transaction/Transaction';
import TransactionSignature from '@bsv/sdk/primitives/TransactionSignature';
// import { BaseChain } from '../tx/BaseChain';
// import type { UTXO, BuildDualFeePoolBaseTxResponse } from '../tx/types';
// import { TripleEndpointPool_1base_tx } from './triple_endpoint_pool_1base_tx';
// import OP from '@bsv/sdk/script/OP';
// import LockingScript from '@bsv/sdk/script/LockingScript';
// import UnlockingScript from '@bsv/sdk/script/UnlockingScript';
// import { fromBase58Check } from '@bsv/sdk/primitives/utils';
import MultiSig from '$lib/script/MULTISIG';
import P2PKH from '../P2PKH';
// import unlock from 'lucide-svelte/icons/unlock';

// 定义 SigHash 常量，与 Go SDK 保持一致
// const SigHash = {
//   SIGHASH_ALL: TransactionSignature.SIGHASH_ALL,
//   FORKID: TransactionSignature.SIGHASH_FORKID
// };

// 定义返回类型
// interface BuildTripleSpendTxResponse {
//   tx: Transaction;
//   clientSignBytes: Buffer;
//   amount: number;
// }

interface TripleSpendTxResponse {
  tx: Transaction;
  clientSignBytes: Buffer;
  amount: number;
}



  



  /**
   * 构建三方多签花费交易
   * @param aTx A方的基础交易
   * @param serverValue 服务器提供的金额
   * @param endHeight 锁定到的区块高度
   * @param serverPublicKey 服务器公钥
   * @param aPrivateKey A方私钥
   * @param bPublicKey B方公钥
   * @returns 完整的交易、客户端签名和金额
   */
  export async function tripleBuildFeePoolSpendTX(
    prevTxId: string,
    serverValue: number,
    endHeight: number,
    serverPublicKey: PublicKey,
    aPrivateKey: PrivateKey,
    bPublicKey: PublicKey,
    feeRate: number,
  ): Promise<TripleSpendTxResponse> {
    try {
      // const prevTxId = aTx.id('hex');
      const aPublicKey = aPrivateKey.toPublicKey();
      const aAddress = aPublicKey.toAddress();
      const bAddress = bPublicKey.toAddress();
      
      // 创建交易对象
      const tx = new Transaction();
      tx.lockTime = endHeight;
      
      // 创建前一个交易的多签锁定脚本
      const prevLockingScript = new MultiSig().lock([serverPublicKey, aPublicKey, bPublicKey], 2);
      const fakeUnlockScript = MultiSig.createFakeSign(2);
      
      // 添加输入（多签 UTXO）
      tx.addInput({
        sourceTXID: prevTxId,
        sourceOutputIndex: 0,
        unlockingScript: fakeUnlockScript,
        sequence: 1, // 设置序列号为1
      });

      tx.inputs[0].sourceTransaction = new Transaction();
      tx.inputs[0].sourceTransaction.outputs = [];
      tx.inputs[0].sourceTransaction.outputs[0] = {
        satoshis: serverValue,
        lockingScript: prevLockingScript
      };
      
      // 创建服务器找零脚本（P2PKH）
      const serverChangeScript = new P2PKH().lock(bPublicKey);
      
      // 添加服务器输出
      tx.addOutput({
        lockingScript: serverChangeScript,
        satoshis: 0, // 初始为0，后续会更新
      });
      
      // 创建客户端找零脚本（P2PKH）
      const clientChangeScript = new P2PKH().lock(aPublicKey);
      
      // 添加客户端输出
      tx.addOutput({
        lockingScript: clientChangeScript,
        satoshis: serverValue, // 初始设置为服务器提供的金额
      });
      
      // 计算交易大小和费用
      const txSize = tx.toBinary().length;
      let fee = Math.floor((txSize / 1000.0) * feeRate);
      
      if (serverValue < fee) {
        throw new Error(`余额不足，需要 ${fee}，拥有 ${serverValue}`);
      }
      
      if (fee === 0) {
        fee = 1; // 最低手续费为 1 satoshi
      }
      
      // 更新客户端输出金额（扣除手续费）
      tx.outputs[1].satoshis = serverValue - fee;

      console.log('------------------------------- BuildOneB success');
      console.log('交易:', tx.toHex());
      
      // 创建 A 方签名
      const SignOne = new MultiSig().signOne(
        tx, 
        0, 
        aPrivateKey, 
        TransactionSignature.SIGHASH_ALL | TransactionSignature.SIGHASH_FORKID
      );

      console.log('A方签名字节:', SignOne.toString('hex'));
      
      return {
        tx: tx,
        clientSignBytes: SignOne,
        amount: serverValue - fee
      };
    } catch (error) {
      console.error('BuildTripleFeePoolSpendTX error:', error);
      throw error;
    }
  }
// }